README.md
# fantasy-football-scoreboard
![I promise to change this picture when I actually build my own](imgs/Scoreboard.jpg)

Display your favourite fantasy football team score on an raspberry pi powered LED matrix. Currently supports 64x32 boards only, and EITHER the Sleeper OR YAHOO fantasy platforms! We have Yahoo!!!

### NFL LED Scoreboard
Hey, I also made an [NFL LED scoreboard](https://github.com/mikemountain/nfl-led-scoreboard), which you really should go check out and star. Because, let's face it, you're gonna hate seeing your fantasy team after a while.

### Credit and inpsiration
This project was inspired by the [nhl-led-scoreboard](https://github.com/riffnshred/nhl-led-scoreboard), who based THEIR project off of the [mlb-led-scoreboard](https://github.com/MLB-LED-Scoreboard/mlb-led-scoreboard). Go check them out, and start watching hockey if you don't already (and baseball too but I love hockey more (go Leafs!)).

### Donate
<a href="https://paypal.me/themikemountain/"><img src="https://github.com/andreostrovsky/donate-with-paypal/blob/master/dark.svg" height="40"></a>

If you enjoyed this project, my NFL project, or if you're just feeling generous, consider buying me a beer. Cheers! :beers: 
You can also PM me on reddit under /u/mikemountain if you need help but don't think it requires an issue!

## Features (v1.0.0)

### UPGRADING TO v1.0.0
There have been some decent changes since v0.0.5. Check the Troubleshooting section if you get any errors running the following steps, but to get to this latest version:
1. Run the command `git stash`. This will take any "local changes" you made (such as to the config, etc) and "store" them so that you can update the local code
2. Run the command `git fetch && git checkout master`. This will tell your scoreboard to look for new updates, and then switch to the main code branch.
3. Run the command `git pull`. This will pull down all the latest changes to your board.
4. Run the command `cp config.json.example config.json`. This will make a copy of the example config to use as your regular config file.
5. Run the command `nano config.json` and then edit in the fields you require (steps for Yahoo and ESPN are further below in the README)
6. After you've set up the config.json, run the board as normal `sudo python main.py` with whatever flags you choose to use.
7. Watch your team lose, try not to cry, cry a lot.

### YAHOO SETUP STUFF
Okay, so Yahoo is gonna be a bit funky. 
1. First, go to https://developer.yahoo.com/ then 'My Apps' and then 'YDN Apps'.
2. On the lefthand panel, click 'Create an App'.
3. Name the app something like "Fantasy Football Scoreboard" in the 'Application Name' block.
4. In the Redirect URL section, just enter https://localhost:8080.
5. Under the 'API Permissions' sections select 'Fantasy Sports' and then make sure that 'Read' is selected.
6. Click create app.
Woo, you've technically created a Yahoo App and you will see two important pieces of information: the Client ID (Consumer Key), and Client Secret(Consumer Secret), which will look like long strings of random letters and numbers.
You will need to enter these two pieces of information into the config, along with your League ID. 
When you run this for the first time, you're going to get some Yahoo stuff happening - a browser window will pop up (if possible), and a link to that browser window will also appear in the console. It'll look similar to something like `https://api.login.yahoo.com/oauth2/request_auth?client_id=CLIENTIDHERE&redirect_uri=oob&language=en-us&response_type=code` and this is normal. If you don't get the browser window popping up, copy and paste this URL into a browser window, and log into Yahoo to authorise the scoreboard. Once you log in, you'll be given a code, which you'll need to type into the console. This is just to authorise the app, and a `token.json` file will be generated into an `./auth` directory so that this doesn't happen every time.

After this, run `sudo pip install yahoo_oauth` so that you have the required library to interact with Yahoo.

I _think_ this should be all the info people need - feel free to reach out if you need more help!

### ESPN SETUP STUFF
You'll need a few pieces of info for ESPN to work - namely some cookies, your league ID and your team ID. You can get your league ID and team ID by just simply looking at the url when you click your team homepage/roster page. In Chrome, you can go to Preferences -> Advanced -> Content Settings -> Cookies -> See all cookies and site data, look for ESPN, and put the "content" of the SWID and the ESPN_S2 cookies in the config, INCLUDING THE {CURLY BRACES} of the SWID.
Note!!! ESPN allows SVGs as team images. There's no easy way for the project to display SVGs, or even convert them to PNGs, so the actual best solution is this:
1. Run the project. If there are SVGs it will download the files, then alert you, and then fail to run properly.
2. Run `sudo apt-get update && apt-get install inkscape`
3. Run `cd logos; find -name "*.svg" -exec sh -c 'sudo inkscape $1 --export-png=${1%.svg}.png' _ {} \;`
If you see `Bitmap saved as:` lines, you should be able to `cd ..` back into the project and then run it again, and it should work. I apologise that it's a bit annoying but it's the way she goes (for now).

### Updating!
Run `bash update.sh` after a pull to make sure you have the necessary packages because I actually don't know a better way to do this.

### Pregame
Currently shows your opponent's avatar, and their name (if it's 12 characters or less, otherwise it won't fit, see the picture at the top of the README). ![nameless and shameless](imgs/no_team_name_preview.jpg) Hoping to incorporate projections in future releases.

### Live scoring updates 
Starting at ~8pm Eastern Thursday, the score will be updated every 20s during game times (Thursday 8pm-12am, Sunday 1pm-12am, Monday 8pm-12am). ![live matchup](imgs/live_matchup.jpg) The colours will change red if a score goes down, and green if a score goes up. ![colour score](imgs/score_changes.jpg) There is also a "big play" notifier of when a team's score goes up by more than 5 points, because that's exciting. The team's score will go gold to show who got the big play. ![big play](imgs/big_play_capture.jpg)

Here's a gif that shows you what this would look like (excuse the shaky hands please, I was updating my testing REST API with one hand and filming with the other). ![score gif](imgs/big_play_and_updates.gif)

I plan to set this so that it only does these checks during actual game times, because there's no real point in checking for game updates on non-game days or during non-game times on gamedays. Eventually, I'd like to only check for score updates if there's a player in the matchup who's playing.

### Postgame
The board will stay in a post-game state until the next week, and will easily disappoint you with a quick glance. Loser is red, winner is green, with LOSS or WIN in between for that extra oomph. ![post game recap](imgs/accurate_postgame.jpg)

### Off season
It displays a message that it's the off season. ![man it's offseason, take a break](imgs/off_season.jpg) You should just turn it off and plan to be heartbroken again next year.

## Roadmap

Future plans include:
* cycle through league scores on off-game times during the week (Post game could cycle through each matchup's result)
* different animations for good plays vs bad plays (nobody wants to see "BIG PLAY" and then see it's your opponent getting the points)
* cycle through multiple teams in multiple leagues so you don't just have to pick your favourite team (although we all have one best league)
* maybe some fun stuff for the draft like who just drafted whom and a countdown clock or something I don't know but it'll be flashy (can't do this yet with current sleeper api)
* analyze your team weaknesses and help with waiver pickups (will not do this)

## Installation
### Hardware Assembly
The [mlb-led-scoreboard guys made a great wiki page to cover the hardware part of the project](https://github.com/MLB-LED-Scoreboard/mlb-led-scoreboard/wiki). There's also this [very handy howchoo page](https://howchoo.com/g/otvjnwy4mji/diy-raspberry-pi-nhl-scoreboard-led-panel) which is what I mainly followed.

### Software Installation
#### Raspbian Distribution
It is recommended you install the Lite version of Raspbian from the [Raspbian Downloads Page](https://www.raspberrypi.org/downloads/raspbian/). This version lacks a GUI, allowing your Pi to dedicate more system resources to drawing the screen.

#### Requirements
You need Git for cloning this repo and PIP for installing the scoreboard software.
```
sudo apt-get update
sudo apt-get install git python-pip
```

#### Installing the software
This installation process might take some time because it will install all the dependencies listed below.

```
git clone --recursive https://github.com/mikemountain/fantasy-football-scoreboard
cd fantasy-football-scoreboard/
sudo chmod +x install.sh
sudo ./install.sh
```
[rpi-rgb-led-matrix](https://github.com/hzeller/rpi-rgb-led-matrix/tree/master/bindings/python#building): The open-source library that allows the Raspberry Pi to render on the LED matrix.

[requests](https://requests.kennethreitz.org/en/master/): To call the API and manipulate the received data.

## Testing & Optimization (IMPORTANT)
If you have used a LED matrix on a raspberry pi before and know how to run it properly, then you can skip this part. 

If you just bought your LED matrix and want to run this software right away, reference the [rpi-rgb-led-matrix library](https://github.com/hzeller/rpi-rgb-led-matrix/). Check out the section that uses the python bindings and run some of their examples on your screen. For sure you will face some issues at first, but don't worry, more than likely there's a solution you can find in their troubleshooting section.
Once you found out how to make it run smoothly, come back here and do what's next.

### Adafruit HAT/bonnet
If you are using any thing from raspberry pi 3+ to the newest versions with an Adafruit HAT or Bonnet, here's what [RiffnShred](https://github.com/riffnshred) did to run his board properly. It seems these are more recommendations than things you 100% absolutely need to do, but are probably beneficial anyway.

* Do the hardware mod found in the [Improving flicker section ](https://github.com/hzeller/rpi-rgb-led-matrix#improving-flicker).
* Disable the on-board sound. You can find how to do it from the [Troubleshooting sections](https://github.com/hzeller/rpi-rgb-led-matrix#troubleshooting)
* From the same section, run the command that remove the bluetooth firmware, unless you use any bluetooth device with your pi.

Finally, here's the command he used.
```
sudo python main.py --led-gpio-mapping=adafruit-hat-pwm --led-brightness=60 --led-slowdown-gpio=2
```

### Configuration
Copy the config.json.example file from the root folder, save it as config.json and set the values you need:

```
platform - set this to either sleeper, yahoo, or espn.

ESPN
league_id           Set this to your ESPN league ID. This is a code that can be found in the URL after leagueId= when you visit your team homepage/roster page.
team_id             Set this to your ESPN team ID. This is a number that can be found in the URL after teamId= when you visit your team homepage/roster page.
swid                In Chrome, go to Preferences -> Advanced -> Content Settings -> Cookies -> See all cookies and site data, look for ESPN, copy the SWID Content data and paste it here INCLUDING THE { BRACES }.
espn_s2             In Chrome, go to Preferences -> Advanced -> Content Settings -> Cookies -> See all cookies and site data, look for ESPN, copy the espn_s2 Content data and paste it here.

SLEEPER
league_id           this value can be found in your Sleeper league's URL: https://sleeper.app/leagues/<league_id>/team
user_id             Run the following command, value will be listed in the output. Replace <username> with the username that you use to login to Sleeper: curl "https://api.sleeper.app/v1/user/<username>"

YAHOO
consumer_key        This will be found in the YDN info that you generate using the Yahoo Setup Stuff section above.
consumer_secret     This will be found in the YDN info that you generate using the Yahoo Setup Stuff section above.
league_id           You should be able to find this info in the URL, I'm not positive where but it should be in there (haven't seen it)
game_code           Don't change this! It corresponds to "NFL" for some reason in the Yahoo API. It changes yearly. This needs to be automated but I have forgotten how I figured this info out because I've been working on this stuff for _way_ too long recently.

year                This corresponds to this year because things will break without this and I haven't gotten around to pulling this info (can't just grab the date info because the season goes into 2022)
debug               Just a flag that prints out more debug info.
```

Now, in a terminal, cd to the fantasy-football-scoreboard folder and run this command. 
```
sudo python main.py 
```
**If you run your screen on an Adafruit HAT or Bonnet, you need to supply this flag.**
```
sudo python main.py --led-gpio-mapping=adafruit-hat
```

### Troubleshooting
* If you run `git fetch` or something and get `error: cannot open .git/FETCH_HEAD: Permission denied` then run the command `sudo chown -R $(whoami) .git/` which will fix the permissions for git. Re-run the failed command.
* If using Yahoo, you may need to run `mkdir auth; sudo chmod 0777 auth/` if you get an `IOError: [Errno 13] Permission denied` error.

### Flags
Use the same flags used in the [rpi-rgb-led-matrix](https://github.com/hzeller/rpi-rgb-led-matrix/) library to configure your screen.
```
--led-rows                Display rows. 16 for 16x32, 32 for 32x32. (Default: 32)
--led-cols                Panel columns. Typically 32 or 64. (Default: 32)
--led-chain               Daisy-chained boards. (Default: 1)
--led-parallel            For Plus-models or RPi2: parallel chains. 1..3. (Default: 1)
--led-pwm-bits            Bits used for PWM. Range 1..11. (Default: 11)
--led-brightness          Sets brightness level. Range: 1..100. (Default: 100)
--led-gpio-mapping        Hardware Mapping: regular, adafruit-hat, adafruit-hat-pwm
--led-scan-mode           Progressive or interlaced scan. 0 = Progressive, 1 = Interlaced. (Default: 1)
--led-pwm-lsb-nanosecond  Base time-unit for the on-time in the lowest significant bit in nanoseconds. (Default: 130)
--led-show-refresh        Shows the current refresh rate of the LED panel.
--led-slowdown-gpio       Slow down writing to GPIO. Range: 0..4. (Default: 1)
--led-no-hardware-pulse   Don't use hardware pin-pulse generation.
--led-rgb-sequence        Switch if your matrix has led colors swapped. (Default: RGB)
--led-pixel-mapper        Apply pixel mappers. e.g Rotate:90, U-mapper
--led-row-addr-type       0 = default; 1 = AB-addressed panels. (Default: 0)
--led-multiplexing        Multiplexing type: 0 = direct; 1 = strip; 2 = checker; 3 = spiral; 4 = Z-strip; 5 = ZnMirrorZStripe; 6 = coreman; 7 = Kaler2Scan; 8 = ZStripeUneven. (Default: 0)
```
There are also flags to set your sleeper league ID and fantasy team ID to use instead of what's in the config (to show other people their teams quickly, for example)
```
--league-id              Set your sleeper league ID
--team-id                Set your fantasy team ID
```

## Licensing
This project uses the GNU General Public License v3.0. If you intend to sell these, the code must remain open source and you at least have to tell your leaguemates how cool I am (please, I need this).



CHANGELOG.md
# Changelog

All notable changes to this project will hopefully be documented in this file.
Still pretty janky.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres as best as I can bother to try 
to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

NEED TO INSTALL YFPY

## [1.0.0] - 2021-09-12

### Added
- ESPN added too!!!!!!

### Changed
- readme
- score update rate changed to every 10 seconds

### Removed

### Fixed
- better way of checking Sleeper scores

## [0.1.0] - 2021-09-04

### Added
- holy shit, Yahoo works
- config example

### Changed
- had to change around the sleeper api file to make it flex with yahoo
- the config stuff
- updated README to include the Yahoo stuff, as well as a way to contact me without making an issue (if needed)

### Removed
- my own config which I never should have committed tbh
- some of the drafting stuff, I never got it to work and there were higher priorities

### Fixed

## [0.0.5] - 2020-10-11

### Added
- added point diff changes because the green/red is nice but I have a goldfish memory, so now I can see how much the scores have changed by

### Changed

### Removed

### Fixed
- there'd be a full minute after boot that it wouldn't check scores, fixed

## [0.0.4 HOTFIX] - 2020-09-12

### Added
- update script to make sure all required packages are installed

### Changed
- added update thing to readme
- commented out more draft stuff, I should just remove it all but I want to work on it eventually

### Removed

### Fixed
- grabbed the bench points by accident, now only grabs regular points

## [0.0.4] - 2020-09-11

### Added

### Changed
- readme
- scores are now getting pulled from different source, updates slower but actually works now so who cares?
- efficient score checking (only checks during game times on thur/sun/mon, can improve this later too)

### Removed
- all the draft stuff now because there's no point now

### Fixed
- technically everything because it didn't actually work before, and now it does. More specifically, I fixed getting the score

## [0.0.3] - 2020-09-09

### Added
- will display the team name instead of username if it's available and fits

### Changed
- readme
- if the week is 0, we'll move into pre-game of week 1 now (some ugly code there)
- removed the player-download until I do something with it
- countdown stuff got moved to a common function
- little visual tweaks

### Removed
- some of the draft stuff because I didn't get it to finish it in time
- took out --team-id option thing in the args because it wasn't working as I intended, gone til fixed

### Fixed
- countdown didn't count down to actual gametime, now it does

## [0.0.2] - 2020-08-14

### Added
- added a changelog I guess
- general robustness, it was pretty janky (still is)
- added a bit of pre-draft/draft stuff (still jank)
- added countdown to kickoff after draft

### Changed
- readme crap
- made a refresh image function to save space

### Removed

### Fixed
- checks for weeks being negative (pre-season, duh)
- avatars weren't getting downloaded
- hopefully logos dir gets created with proper permissions now


utils.py
from __future__ import division
from builtins import range
# from past.utils import old_div
from rgbmatrix import RGBMatrixOptions, graphics
import collections
import argparse
import os
import debug
import datetime
from tzlocal import get_localzone
import pytz

# get local timezone
local_tz = get_localzone()

def get_file(path):
  dir = os.path.dirname(__file__)
  return os.path.join(dir, path)

def center_text(text_width,center_pos):
  # return abs(center_pos - (old_div(text_width, 2)))
  return abs(center_pos - (text_width / 2))

def split_string(string, num_chars):
  return [(string[i:i + num_chars]).strip() for i in range(0, len(string), num_chars)]

def args():
  parser = argparse.ArgumentParser()

  # Options for the rpi-rgb-led-matrix library
  parser.add_argument("--led-rows", action="store", help="Display rows. 16 for 16x32, 32 for 32x32. (Default: 32)", default=32, type=int)
  parser.add_argument("--led-cols", action="store", help="Panel columns. Typically 32 or 64. (Default: 64)", default=64, type=int)
  parser.add_argument("--led-chain", action="store", help="Daisy-chained boards. (Default: 1)", default=1, type=int)
  parser.add_argument("--led-parallel", action="store", help="For Plus-models or RPi2: parallel chains. 1..3. (Default: 1)", default=1, type=int)
  parser.add_argument("--led-pwm-bits", action="store", help="Bits used for PWM. Range 1..11. (Default: 11)", default=11, type=int)
  parser.add_argument("--led-brightness", action="store", help="Sets brightness level. Range: 1..100. (Default: 100)", default=100, type=int)
  parser.add_argument("--led-gpio-mapping", help="Hardware Mapping: regular, adafruit-hat, adafruit-hat-pwm" , choices=['regular', 'adafruit-hat', 'adafruit-hat-pwm'], type=str)
  parser.add_argument("--led-scan-mode", action="store", help="Progressive or interlaced scan. 0 = Progressive, 1 = Interlaced. (Default: 1)", default=1, choices=list(range(2)), type=int)
  parser.add_argument("--led-pwm-lsb-nanoseconds", action="store", help="Base time-unit for the on-time in the lowest significant bit in nanoseconds. (Default: 130)", default=130, type=int)
  parser.add_argument("--led-show-refresh", action="store_true", help="Shows the current refresh rate of the LED panel.")
  parser.add_argument("--led-slowdown-gpio", action="store", help="Slow down writing to GPIO. Range: 0..4. (Default: 1)", choices=list(range(5)), type=int)
  parser.add_argument("--led-no-hardware-pulse", action="store", help="Don't use hardware pin-pulse generation.")
  parser.add_argument("--led-rgb-sequence", action="store", help="Switch if your matrix has led colors swapped. (Default: RGB)", default="RGB", type=str)
  parser.add_argument("--led-pixel-mapper", action="store", help="Apply pixel mappers. e.g \"Rotate:90\"", default="", type=str)
  parser.add_argument("--led-row-addr-type", action="store", help="0 = default; 1 = AB-addressed panels. (Default: 0)", default=0, type=int, choices=[0,1])
  parser.add_argument("--led-multiplexing", action="store", help="Multiplexing type: 0 = direct; 1 = strip; 2 = checker; 3 = spiral; 4 = Z-strip; 5 = ZnMirrorZStripe; 6 = coreman; 7 = Kaler2Scan; 8 = ZStripeUneven. (Default: 0)", default=0, type=int)

  # User Options
  # these are broken atm
  # parser.add_argument("--user-id", action="store", help="ID of your fantasy user. Change the default in the config.json", type=int)
  # parser.add_argument("--league-id", action="store", help="ID of your sleeper fantasy league. Change the default in the config.json,", type=int)

  return parser.parse_args()

def led_matrix_options(args):
  options = RGBMatrixOptions()

  if args.led_gpio_mapping != None:
    options.hardware_mapping = args.led_gpio_mapping

  options.rows = args.led_rows
  options.cols = args.led_cols
  options.chain_length = args.led_chain
  options.parallel = args.led_parallel
  options.row_address_type = args.led_row_addr_type
  options.multiplexing = args.led_multiplexing
  options.pwm_bits = args.led_pwm_bits
  options.brightness = args.led_brightness
  options.pwm_lsb_nanoseconds = args.led_pwm_lsb_nanoseconds
  options.led_rgb_sequence = args.led_rgb_sequence
  try:
    options.pixel_mapper_config = args.led_pixel_mapper
  except AttributeError:
    debug.warning("Your compiled RGB Matrix Library is out of date.")
    debug.warning("The --led-pixel-mapper argument will not work until it is updated.")

  if args.led_show_refresh:
    options.show_refresh_rate = 1

  if args.led_slowdown_gpio != None:
    options.gpio_slowdown = args.led_slowdown_gpio

  if args.led_no_hardware_pulse:
    options.disable_hardware_pulsing = True

  return options


def deep_update(source, overrides):
    """Update a nested dictionary or similar mapping.
    Modify ``source`` in place.
    """
    for key, value in list(overrides.items()):
        if isinstance(value, collections.Mapping) and value:
            returned = deep_update(source.get(key, {}), value)
            source[key] = returned
        else:
            source[key] = overrides[key]
    return source

def convert_time(utc_dt):

  local_dt = datetime.datetime.strptime(utc_dt, '%Y-%m-%dT%H:%M:%SZ').replace(tzinfo=pytz.utc).astimezone(local_tz)
  return local_tz.normalize(local_dt)  # .normalize might be unnecessary